pymc/model.py

```diff
@@ -718,50 +718,53 @@ def logp_dlogp_function(self, grad_vars=None, tempered=False, **kwargs):
                 # Convert random variables into their log-likelihood inputs and
                 # apply their transforms, if any
                 potentials, _ = rvs_to_value_vars(self.potentials, apply_transforms=True)
+                potentials_logp = at.sum([at.sum(factor) for factor in potentials])
 
-                free_RVs_logp = at.sum(
-                    [at.sum(logpt(var, self.rvs_to_values.get(var, None))) for var in self.free_RVs]
-                    + list(potentials)
-                )
-                observed_RVs_logp = at.sum(
-                    [at.sum(logpt(obs, obs.tag.observations)) for obs in self.observed_RVs]
-                )
+                # TODO: Here potentials are added to the free_RVs but this is
+                #  the opposite of  what varlogpt and datalogpt do. Otherwise we could
+                #  just call self.datalogpt below
+                free_RVs_logp = self.varlogpt + potentials_logp
+
+                obs_values = {}
+                for obs in self.observed_RVs:
+                    obs_values[obs] = obs.tag.observations
+                if obs_values:
+                    observed_RVs_logp = logpt(self.observed_RVs, obs_values)
 
             costs = [free_RVs_logp, observed_RVs_logp]
         else:
             costs = [self.logpt]
 
         input_vars = {i for i in graph_inputs(costs) if not isinstance(i, Constant)}
         extra_vars = [self.rvs_to_values.get(var, var) for var in self.free_RVs]
+        ip = self.recompute_initial_point(0)
         extra_vars_and_values = {
-            var: self.initial_point[var.name]
-            for var in extra_vars
-            if var in input_vars and var not in grad_vars
+            var: ip[var.name] for var in extra_vars if var in input_vars and var not in grad_vars
         }
         return ValueGradFunction(costs, grad_vars, extra_vars_and_values, **kwargs)
 
     @property
     def logpt(self):
         """Aesara scalar of log-probability of the model"""
+        logp_var = 0
 
         rv_values = {}
         for var in self.free_RVs:
             rv_values[var] = self.rvs_to_values.get(var, None)
-        rv_factors = logpt(self.free_RVs, rv_values)
+        if rv_values:
+            logp_var += logpt(self.free_RVs, rv_values)
 
         obs_values = {}
         for obs in self.observed_RVs:
             obs_values[obs] = obs.tag.observations
-        obs_factors = logpt(self.observed_RVs, obs_values)
+        if obs_values:
+            logp_var += logpt(self.observed_RVs, obs_values)
```

Comment:
@kc611 Do you remember if there was a reason why we need two separate calls to `logpt` and why can't we just pass `rv_values` + `obs_values` once?

Reply:
I guess in some cases there were overlapping keys in those value dictionaries.

Reply:
That's not an issue for dictionaries, right?